WebRTC types should be serialized using CoreIPC primitive Strings instead of raw bytes

gavin-apple opened the pull request:
#### 6ed6b6a4c121b050e54bedea8db4a811aa6fa323
<pre>
WebRTC types should be serialized using CoreIPC primitive Strings instead of raw bytes
<a href="https://bugs.webkit.org/show_bug.cgi?id=303055">https://bugs.webkit.org/show_bug.cgi?id=303055</a>
<a href="https://rdar.apple.com/165358157">rdar://165358157</a>

Reviewed by Youenn Fablet.

Ports WebRTC strings from ODTs to native serializable
CoreIPC primitive Strings.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
(WebKit::NetworkRTCMonitor::sortNetworks):
* Source/WebKit/Scripts/webkit/opaque_ipc_types.tracking.in:
* Source/WebKit/Shared/RTCNetwork.cpp:
(WebKit::RTCNetwork::RTCNetwork):
(WebKit::RTCNetwork::value const):
(WebKit::WebRTCNetwork::SocketAddress::rtcAddress const):
(WebKit::WebRTCNetwork::SocketAddress::SocketAddress):
(WebKit::WebRTCNetwork::ipAddress): Deleted.
* Source/WebKit/Shared/RTCNetwork.h:
(WebKit::WebRTCNetwork::SocketAddress::SocketAddress):
* Source/WebKit/Shared/RTCNetwork.serialization.in:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::networksChanged):

Canonical link: <a href="https://commits.webkit.org/303624@main">https://commits.webkit.org/303624@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc6d58e3108c652eabb5a7f1ee334aa17098d100

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132980 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5481 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44096 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140516 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85011 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/94101906-aac2-42a1-ba32-25ff302591c6) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6008 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5344 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101688 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69016 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d3a54ac2-d300-448f-9366-3e5ea7ebfc62) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135926 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4226 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119174 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82487 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/54b40589-28c5-4781-93cb-fb1584964123) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/132329 "Passed tests") | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4109 "Found unexpected failure with change (failure)") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1669 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83750 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113171 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143169 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5150 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37873 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110064 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5232 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4413 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110244 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27960 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3949 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115436 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58721 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5204 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33787 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5045 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68656 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5294 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5162 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jon-b-apple commented on Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
```diff
@@ -22,6 +22,8 @@
 
 webkit_platform_headers: "StreamConnectionEncoder.h"
 
+[Alias=class WTF::String, Validator='result->is8Bit()'] alias Latin1String;
```
There's an implicit introduction of `result` in this validator that was unexpected. Previously I think we only ever validated fields, not whole types. Would it be better to call the validator as a method and shorten it to `is8Bit()`?

Also, I think I would prefer the structure to be:
`[Validator='is8Bit()'] alias Latin1String = class WTFString;`

gavin-apple replied:
The current structure mirrors our existing Aliasing - primarily used for templating. I agree with the implicit result. Do we really want to limit to just methods?

jon-b-apple commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -633,7 +652,7 @@ def generate_forward_declarations(serialized_types, serialized_enums, additional
         else:
             serialized_enums_by_namespace[enum.namespace] = [enum, ]
-    for type in serialized_types:
+    for type in serialized_types + alias_types:
```
Consider using `itertools.chain` to avoid creating a new list to loop over here

jon-b-apple commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -1053,7 +1097,15 @@ def construct_type(type, specialization, indentation):
+def generate_type_validation(type):
+    result = []
+    if type.type_validator:
+        result.append(f'    if (!' + type.type_validator + ')')
```
The validator expression should go in parens to make sure the `!` doesn't affect it, like we do when validating members

jon-b-apple commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -1214,14 +1276,14 @@ def generate_impl(serialized_types, serialized_enums, headers, generating_webkit
     result.append('')
 
-    for type in serialized_types:
+    for type in serialized_types + alias_types:
```
Use `itertools.chain` here too

jon-b-apple commented on Source/WebKit/Shared/RTCNetwork.cpp:
```diff
@@ -45,7 +45,13 @@ static_assert
-RTCNetwork::RTCNetwork(Vector<char>&& name, Vector<char>&& description, IPAddress prefix, int prefixLength, int type, uint16_t id, int preference, bool active, bool ignored, int scopeID, Vector<InterfaceAddress>&& ips)
+static absl::string_view toStringView(const String& string)
+{
+    ASSERT(string.is8Bit());
```
Should we enforce this more strictly in release? Also this requirement isn't obvious from the name, in future I could imagine someone wanting to call this with a 16-bit string that might not be going over IPC and getting a crash. We could call it `to8BitStringView` and fail with an empty view if String was 16-bit?

gavin-apple committed:
WebRTC types should be serialized using CoreIPC primitive Strings instead of raw bytes
https://bugs.webkit.org/show_bug.cgi?id=303055
rdar://165358157

Reviewed by Youenn Fablet.

Ports WebRTC strings from ODTs to native serializable
CoreIPC primitive Strings.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
(WebKit::NetworkRTCMonitor::sortNetworks):
* Source/WebKit/Scripts/webkit/opaque_ipc_types.tracking.in:
* Source/WebKit/Shared/RTCNetwork.cpp:
(WebKit::RTCNetwork::RTCNetwork):
(WebKit::RTCNetwork::value const):
(WebKit::WebRTCNetwork::SocketAddress::rtcAddress const):
(WebKit::WebRTCNetwork::SocketAddress::SocketAddress):
(WebKit::WebRTCNetwork::ipAddress): Deleted.
* Source/WebKit/Shared/RTCNetwork.h:
(WebKit::WebRTCNetwork::SocketAddress::SocketAddress):
* Source/WebKit/Shared/RTCNetwork.serialization.in:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::networksChanged):

Canonical link: https://commits.webkit.org/303624@main
